### ISSUE TITLE
Not able to reproduce the same results with the C++ API. I get different output result.

### ISSUE THREAD

**Issue by Con-Mi** [pytorch/pytorch#14273]
## Cpp Label

I have trained a segmentation model in Python with 4 Gpus. When I try the model in the Python environment I get the results I expect.
I export then the Python model and make it readable in C++ and use the libtorch header files to read it into the C++ model, but when I run the forward function, it does not seem to produce the same results.

1. Check this gist where I have all the files:
https://gist.github.com/Con-Mi/4d92af62adb784a5353ff7cf19d6d099

2. Download the image 0010_.png and the binary var_dense_linknet_384_sgd_bce_20epchs.pt
from the link below
https://www.dropbox.com/sh/jni3lpsk68iu0fd/AACudbOBnzufHy0ttHpeTLDsa?dl=0

3. Run python convert_bin.py
(I am using PyTorch 1.0(Python 3.6) to do this you need this version.)
It will save the jit model that you need to load into C++ application.

4. Build the project with CMake, just by running cmake ..  , in the build folder.

5. Run the program with: ./testing 
After that they will be windows of the original image and the output.

6. Run the python program python_prediction.py in order to compare the results

Dependencies: 
   - OpenCV 3.4.3-dev in C++. Higher version will work as well. Not a particular preference. (Just change it in the CmakeLists.txt)
   - libtorch
  - Python Libraries:
          - Matplotlib
          - Seaborn (You can skip this if you want. Just remove the lines in the code.)
          - PyTorch 1.0 and torchvision
          - Numpy 
          - Opencv 3.4

**Comment by Con-Mi** [comment]
@bhack I have posted them in the dropbox folder as python_out.png and cpp_out.png. 
Check if you find them.

**Comment by Con-Mi** [comment]
@goldsborough Is it something I am doing wrong in the C++ part?

**Comment by bhack** [comment]
@Con-Mi Do you need to permute https://gist.github.com/Con-Mi/4d92af62adb784a5353ff7cf19d6d099#file-cpp_model-cc-L16? 
Opencv MATs are NHWC with C BGR.

**Comment by Con-Mi** [comment]
cv::IMREAD_UNCHANGED reads the image as it is.
https://docs.opencv.org/3.1.0/d4/da8/group__imgcodecs.html#gga61d9b0126a3e57d9277ac48327799c80aeddd67043ed0df14f9d9a4e66d2b0708

**Comment by tholzmann** [comment]
I have a similar problem: We trained a semantic segmentation model in python. When doing inference with it in Python and C++, the results are slightly different (see images below). Also on other images, the difference is not very big, but enough that it is visible.

I'm using exactly the same input data and applying exactly the same pre/post-processing.

I am using the latest nightly build (but I tried also latest stable) for inference with C++, running on Windows 10 with Visual Studio 2019.
Training was done on Ubuntu 18.04 with PyTorch 1.1.

Does someone know a possible reason for this problem? Might this be a bug in the C++ API?

Best,
Thomas

Result of Python:
![result_python](https://user-images.githubusercontent.com/9862284/60729958-440ff600-9f44-11e9-93af-6639cb5c47af.png)

Result of C++ LibTorch:
![result_c++](https://user-images.githubusercontent.com/9862284/60730032-76215800-9f44-11e9-9645-1392979b4e42.png)

**Comment by tholzmann** [comment]
Update: I found the error and now I get exactly the same results. Simple error on our side. 

Sorry for stealing your time.

Best,
Thomas

**Comment by Con-Mi** [comment]
I have made a working solution and I post it here, in case anyone else arrives again on the same issue.
First you need to consider reading an image with OpenCV in C++. When using
```
cv::Mat img = cv::imread(FILENAME_, cv::IMREAD_UNCHANGED);
```
it will read the image with BGR channels and you need to switch it back to RGB.
So you need something like
```
cv::cvtColor( img, img, cv::COLOR_BGR2RGB );
```

Second you need to convert the values of the image as read in OpenCV between [0, 1].
You can do the following:
```
img.convertTo( img, CV_32FC3, 1/255.0 );
```

Next, when you convert an image to tensor image in using toch C++ API, you need to permute the tensor, so it reads as BCHW (stands for batch, channel, height, width) 
You can do this like this:
```
at::Tensor tensorImage = torch::from_blob(img.data, { 1, img.rows, img.cols, 3 }, at::kFloat);
tensorImage = tensorImage.permute({0, 3, 1, 2});
```
Last step before you push it to the model you need to normalize the image, if you are using a pretrained model from the library. You do this per channel.
You can do the following:
```
tensorImage[0][0] = tensorImage[0][0].sub(0.485).div(
tensorImage[0][1] = tensorImage[0][1].sub(0.456).div(
tensorImage[0][2] = tensorImage[0][2].sub(0.406).div(0.225);
```

Now posting full working code in a Python Script and the corresponding C++ code.
Python:
```
import torch
from torchvision import models, transforms
import cv2


_FILENAME = "../images/goldfish.jpg"

model = models.mnasnet0_5(pretrained=True)
model.eval()

tfm = transforms.Compose([ 
        transforms.ToTensor(), 
        transforms.Normalize( mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225] ) 
    ])

image = cv2.imread(_IMAGE_FILENAME)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image = cv2.resize(image, (224, 224))

image = tfm(image)
image = image.unsqueeze(dim=0)

output = model(image)
output = torch.softmax(output, 1)

prob_value, index = torch.max(output, 1)

print("Probability Value: ")
print(prob_value)
print("ImageNet Index: ")
print(index)
```

C++ code:
```
#include <iostream>
#include <vector>
#include <string>

#include <torch/torch.h>
#include <torch/script.h>

#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc.hpp>


int main() {

    const cv::String _IMAGE_FILENAME = "../images/goldfish.jpg";
    const std::string _MODEL_JIT_FILENAME= "../jit_models_bin/traced_mnasnet0_5.pt";

    cv::Mat img = cv::imread( _IMAGE_FILENAME );
    cv::cvtColor( img, img, cv::COLOR_BGR2RGB );
    cv::Size rsz = { 224, 224 };

    cv::resize( img, img, rsz, 0, 0, cv::INTER_LINEAR );
    img.convertTo( img, CV_32FC3, 1/255.0 );

    at::Tensor tensorImage = torch::from_blob(img.data, { 1, img.rows, img.cols, 3 }, at::kFloat);
    tensorImage = tensorImage.permute({0, 3, 1, 2});

    //  Normalize data
    tensorImage[0][0] = tensorImage[0][0].sub(0.485).div(0.229);
    tensorImage[0][1] = tensorImage[0][1].sub(0.456).div(0.224);
    tensorImage[0][2] = tensorImage[0][2].sub(0.406).div(0.225);

    std::vector<torch::jit::IValue> input;
    input.push_back(tensorImage);

    torch::jit::script::Module model = torch::jit::load( _MODEL_JIT_FILENAME );

    at::Tensor output = torch::softmax(model.forward(input).toTensor(), 1);

    std::tuple<at::Tensor, at::Tensor> result = torch::max(output, 1);

    std::cout << "Probability Value: " << std::endl;
    std::cout << std::get<0>(result) << std::endl;
    std::cout << "ImageNet Index" << std::endl;
    std::cout << std::get<1>(result) << std::endl;

    return 0;
}
```

You should get the following results for both:
Probability Value: 0.991
ImageNet Index: 1

I used the following library versions:
PyTorch: 1.3
OpenCV: 4.1
TorchVision: 0.4
